src/mapport.cpp

```diff
@@ -0,0 +1,342 @@
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                ret = false;
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+                break;
+            } else {
+                ret = true;
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+        g_mapport_interrupt.reset();
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+
+    return ret;
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
+{
+    bool ok;
+    do {
+        ok = false;
+
+#ifdef USE_UPNP
+        // High priority protocol.
+        if (g_mapport_enabled_protos & MapPortProtoFlag::UPNP) {
+            g_mapport_current_proto = MapPortProtoFlag::UPNP;
+            ok = ProcessUpnp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_UPNP
+
+#ifdef USE_NATPMP
+        // Low priority protocol.
+        if (g_mapport_enabled_protos & MapPortProtoFlag::NAT_PMP) {
+            g_mapport_current_proto = MapPortProtoFlag::NAT_PMP;
+            ok = ProcessNatpmp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_NATPMP
+
+        g_mapport_current_proto = MapPortProtoFlag::NONE;
+        if (g_mapport_enabled_protos == MapPortProtoFlag::NONE) {
+            return;
+        }
+
+    } while (ok || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD));
+}
+
+void StartThreadMapPort()
+{
+    if (!g_mapport_thread.joinable()) {
+        assert(!g_mapport_interrupt);
+        g_mapport_thread = std::thread(std::bind(&TraceThread<void (*)()>, "mapport", &ThreadMapPort));
+    }
+}
+
+static void DispatchMapPort()
+{
+    if (g_mapport_current_proto == MapPortProtoFlag::NONE && g_mapport_enabled_protos == MapPortProtoFlag::NONE) {
+        return;
+    }
+
+    if (g_mapport_current_proto == MapPortProtoFlag::NONE && g_mapport_enabled_protos != MapPortProtoFlag::NONE) {
+        StartThreadMapPort();
+        return;
+    }
+
+    if (g_mapport_current_proto != MapPortProtoFlag::NONE && g_mapport_enabled_protos == MapPortProtoFlag::NONE) {
+        InterruptMapPort();
+        StopMapPort();
+        return;
+    }
+
+    if (g_mapport_enabled_protos & g_mapport_current_proto) {
+        // Enabling another protocol does not cause switching from the currently used one.
+        return;
+    }
+
+    assert(g_mapport_thread.joinable());
+    assert(!g_mapport_interrupt);
```

Comment:
Couldn't this be true if the user is impossibly fast?

Reply:
I don't think so due to the `CThreadInterrupt` class implementation.